megamek/src/megamek/common/pathfinder/AeroGroundPathFinder.java

```diff
@@ -0,0 +1,647 @@
+             * by ending prematurely while preserving already computed results.
+             */
+
+            candidates = null;
+            //candidates = new PriorityQueue<E>(100, comparator);
+            e.printStackTrace();
+            System.err.println("Not enough memory to analyse all options."//$NON-NLS-1$
+                    + " Try setting time limit to lower value, or "//$NON-NLS-1$
+                    + "increase java memory limit.");//$NON-NLS-1$
+        } catch(Exception e) {
+            e.printStackTrace(); //do something, don't just swallow the exception, good lord
+        }
+    }
+    
+    /**
+     * Gets the shortest of all paths generated so far that explicitly goes off-board.
+     */
+    public MovePath getShortestOffBoardPath() {
+        return offBoardFilter.getShortestPath();
+    }
+    
+    public static AeroGroundPathFinder getInstance(IGame game) {
+        AeroGroundPathFinder apf = new AeroGroundPathFinder(
+                                        new MovePathFinder.MovePathDestinationMap(),
+                                        new ShortestPathFinder.AeroMovePathRelaxer(),
+                                        new NextStepsAeroGroundAdjacencyMap(MoveStepType.FORWARDS),
+                                        new MovePathFinder.MovePathVelocityCostComparator(),
+                                        game);
+        
+        apf.addFilter(new MovePathVelocityFilter());
+        apf.addFilter(new MovePathLegalityFilter(game));
+        return apf;
+    }
+    
+    /**
+     * Worker function that indicates if there is an enemy aircraft on the board for the current path.
+     * Assumes we are on a ground map
+     * @param path the current path we're evaluating
+     * @return Whether or not there's an enemy airborne entity on the board.
+     */
+    private static boolean airborneEnemiesOnBoard(Entity mover) {
+        for(Iterator<Entity> entityIter = mover.getGame().getAllEnemyEntities((Entity) mover); entityIter.hasNext();) {
+            Entity currentEntity = entityIter.next();
+            if(currentEntity.isAero()) {
+                return true;
+            }
+        }
+        
+        return false;
+    }
+    
+    /**
+     * Helper function to calculate the maximum thrust we should use for a particular aircraft
+     * We limit ourselves to the lowest of "safe thrust" and "structural integrity"...
+     *  ... or something a little lower to prevent using gigabyte after gigabyte of memory on high-thrust paths
+     *  that don't do anything useful.
+     * @param aero The aero entity for which to calculate max thrust.
+     * @return The max thrust.
+     */
+    private static int calculateMaxThrust(IAero aero) {
+        int maxThrust = Math.min(aero.getCurrentThrust(), aero.getSI());    // we should only thrust up to our SI
+        //maxThrust = Math.min(maxThrust, artificialCap); // lol just kidding, you ain't going nowhere fast bro
+        return maxThrust;
+    }
```

Comment:
Should a limit based on board size be considered?  There's no point in using more than 1 or 2 thrust when you're on a 30x30 map board.

Reply:
Agreed, though the limit should be on velocity, not thrust. I've placed an artificial cap of 3 on velocity for now due to performance limitations (4 velocity tends to generate upwards of 6000 possible paths on sufficiently large maps, which takes a prohibitive amount of time to process, with a questionable payoff).